File: box-and-box/ports/python/epistemic.py
"""epistemic.py - Epistemic Arithmetic, faithful Python port (v0.6)

Knowledge and graded belief. A model is a universe of possible worlds plus, per
agent, an accessibility relation. Knowledge = truth in ALL accessible worlds.
  S5   (equivalence relation) -> KNOWLEDGE: factive (Kphi->phi), introspective.
  KD45 (serial+transitive+euclidean, NOT reflexive) -> BELIEF: consistent and
       introspective but NOT factive.
Learning is public announcement. Laws E1-E8, EB1-EB3. Mirrors epistemic.mjs.

A world is a dict; access[agent] is a callable world -> list-of-worlds. World
identity is by object identity (mirroring JS indexOf), so worlds must be the same
objects shared across the access closures - exactly as constructed in laws.py.
"""


def atom(name, pred):
    return {"t": "atom", "name": name, "pred": pred}


def not_(a):
    return {"t": "not", "a": a}


def and_(a, b):
    return {"t": "and", "a": a, "b": b}


def or_(a, b):
    return {"t": "or", "a": a, "b": b}


def implies(a, b):
    return {"t": "implies", "a": a, "b": b}


def holds(f, w):
    t = f["t"]
    if t == "atom":
        return bool(f["pred"](w))
    if t == "not":
        return not holds(f["a"], w)
    if t == "and":
        return holds(f["a"], w) and holds(f["b"], w)
    if t == "or":
        return holds(f["a"], w) or holds(f["b"], w)
    if t == "implies":
        return (not holds(f["a"], w)) or holds(f["b"], w)
    return False


def Model(worlds, actual, access):
    return {"worlds": worlds, "actual": actual, "access": access}


def knows_at(model, agent, w, f):
    acc = model["access"][agent](w)
    return len(acc) > 0 and all(holds(f, u) for u in acc)


def possible_at(model, agent, w, f):
    return any(holds(f, u) for u in model["access"][agent](w))  # not K not f


def believes_at(model, agent, w, f, theta=0.5):
    acc = model["access"][agent](w)
    if len(acc) == 0:
        return False
    return sum(1 for u in acc if holds(f, u)) / len(acc) >= theta


def knows(model, agent, f):
    return knows_at(model, agent, model["actual"], f)


def believes(model, agent, f, theta=0.5):
    return believes_at(model, agent, model["actual"], f, theta)


def knows_it_doesnt_know(model, agent, f):
    """The known-unknown (K not K phi): the kappa / 'deliberate' signal."""
    acc = model["access"][agent](model["actual"])
    return len(acc) > 0 and all(not knows_at(model, agent, u, f) for u in acc)


def route(model, agent, f):
    """Epistemic routing: act on what you know, deliberate on a detected gap."""
    if knows(model, agent, f):
        return "act"
    if knows_it_doesnt_know(model, agent, f):
        return "deliberate"
    return "uncertain"


def _identity_filter(seq, keep_objs):
    keep_ids = set(id(x) for x in keep_objs)
    return [u for u in seq if id(u) in keep_ids]


def announce(model, psi):
    """Learning = truthful public announcement: keep only worlds where psi holds."""
    worlds = [w for w in model["worlds"] if holds(psi, w)]
    old_access = model["access"]

    def make(a):
        return lambda w: _identity_filter(old_access[a](w), worlds)

    access = {a: make(a) for a in old_access.keys()}
    return Model(worlds, model["actual"], access)


def everyone(model, agents, f):
    return all(knows(model, a, f) for a in agents)


def common(model, agents, f):
    """Common knowledge: f holds in every world reachable from actual via union of access."""
    reach = [model["actual"]]
    reach_ids = {id(model["actual"])}
    stack = [model["actual"]]
    while stack:
        w = stack.pop()
        for a in agents:
            for u in model["access"][a](w):
                if id(u) not in reach_ids:
                    reach_ids.add(id(u))
                    reach.append(u)
                    stack.append(u)
    return all(holds(f, u) for u in reach)


def distributed(model, agents, f):
    """Distributed knowledge: pool information by intersecting accessible sets."""
    sets = [model["access"][a](model["actual"]) for a in agents]
    first = sets[0]
    inter = []
    for w in first:
        if all(any(id(w) == id(x) for x in s) for s in sets):
            inter.append(w)
    return len(inter) > 0 and all(holds(f, u) for u in inter)

File: box-and-box/ports/go/epistemic.go
package main

// epistemic.go — Epistemic Arithmetic / S5 & KD45 modal logic (port of epistemic.mjs).
// Laws E1–E8, EB1–EB3. Worlds use pointer identity (mirrors JS object identity).

// World is a valuation of atoms. Pointers give identity (like JS objects).
type World = map[string]bool

// EFormula is a propositional formula over worlds.
type EFormula struct {
	T    string // atom,not,and,or,implies
	Name string
	Pred func(w *World) bool
	A    *EFormula
	B    *EFormula
}

func eAtomF(name string, pred func(w *World) bool) *EFormula {
	return &EFormula{T: "atom", Name: name, Pred: pred}
}
func eNot(a *EFormula) *EFormula        { return &EFormula{T: "not", A: a} }
func eAnd(a, b *EFormula) *EFormula     { return &EFormula{T: "and", A: a, B: b} }
func eOr(a, b *EFormula) *EFormula      { return &EFormula{T: "or", A: a, B: b} }
func eImplies(a, b *EFormula) *EFormula { return &EFormula{T: "implies", A: a, B: b} }

func eHolds(f *EFormula, w *World) bool {
	switch f.T {
	case "atom":
		return f.Pred(w)
	case "not":
		return !eHolds(f.A, w)
	case "and":
		return eHolds(f.A, w) && eHolds(f.B, w)
	case "or":
		return eHolds(f.A, w) || eHolds(f.B, w)
	case "implies":
		return !eHolds(f.A, w) || eHolds(f.B, w)
	default:
		return false
	}
}

// EModel : worlds + actual world + per-agent accessibility.
type EModel struct {
	Worlds []*World
	Actual *World
	Access map[string]func(w *World) []*World
}

func knowsAt(m EModel, agent string, w *World, f *EFormula) bool {
	acc := m.Access[agent](w)
	if len(acc) == 0 {
		return false
	}
	for _, u := range acc {
		if !eHolds(f, u) {
			return false
		}
	}
	return true
}

func possibleAt(m EModel, agent string, w *World, f *EFormula) bool {
	for _, u := range m.Access[agent](w) {
		if eHolds(f, u) {
			return true
		}
	}
	return false
}

func believesAt(m EModel, agent string, w *World, f *EFormula, theta float64) bool {
	acc := m.Access[agent](w)
	if len(acc) == 0 {
		return false
	}
	cnt := 0
	for _, u := range acc {
		if eHolds(f, u) {
			cnt++
		}
	}
	return float64(cnt)/float64(len(acc)) >= theta
}

func knows(m EModel, agent string, f *EFormula) bool {
	return knowsAt(m, agent, m.Actual, f)
}
func believes(m EModel, agent string, f *EFormula, theta float64) bool {
	return believesAt(m, agent, m.Actual, f, theta)
}

func knowsItDoesntKnow(m EModel, agent string, f *EFormula) bool {
	acc := m.Access[agent](m.Actual)
	if len(acc) == 0 {
		return false
	}
	for _, u := range acc {
		if knowsAt(m, agent, u, f) {
			return false
		}
	}
	return true
}

func route(m EModel, agent string, f *EFormula) string {
	if knows(m, agent, f) {
		return "act"
	}
	if knowsItDoesntKnow(m, agent, f) {
		return "deliberate"
	}
	return "uncertain"
}

// announce : truthful public announcement — keep only worlds where ψ holds.
func announce(m EModel, psi *EFormula) EModel {
	worlds := []*World{}
	keep := map[*World]bool{}
	for _, w := range m.Worlds {
		if eHolds(psi, w) {
			worlds = append(worlds, w)
			keep[w] = true
		}
	}
	access := map[string]func(w *World) []*World{}
	for a := range m.Access {
		old := m.Access[a]
		access[a] = func(w *World) []*World {
			out := []*World{}
			for _, u := range old(w) {
				if keep[u] {
					out = append(out, u)
				}
			}
			return out
		}
	}
	return EModel{Worlds: worlds, Actual: m.Actual, Access: access}
}

func everyone(m EModel, agents []string, f *EFormula) bool {
	for _, a := range agents {
		if !knows(m, a, f) {
			return false
		}
	}
	return true
}

// common : f holds in every world reachable from actual via the union of agents' access.
func common(m EModel, agents []string, f *EFormula) bool {
	reach := map[*World]bool{m.Actual: true}
	stack := []*World{m.Actual}
	for len(stack) > 0 {
		w := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		for _, a := range agents {
			for _, u := range m.Access[a](w) {
				if !reach[u] {
					reach[u] = true
					stack = append(stack, u)
				}
			}
		}
	}
	for u := range reach {
		if !eHolds(f, u) {
			return false
		}
	}
	return true
}

// distributed : pool information by intersecting accessible sets.
func distributed(m EModel, agents []string, f *EFormula) bool {
	if len(agents) == 0 {
		return false
	}
	sets := make([]map[*World]bool, len(agents))
	for i, a := range agents {
		s := map[*World]bool{}
		for _, w := range m.Access[a](m.Actual) {
			s[w] = true
		}
		sets[i] = s
	}
	inter := []*World{}
	for w := range sets[0] {
		all := true
		for _, s := range sets {
			if !s[w] {
				all = false
				break
			}
		}
		if all {
			inter = append(inter, w)
		}
	}
	if len(inter) == 0 {
		return false
	}
	for _, u := range inter {
		if !eHolds(f, u) {
			return false
		}
	}
	return true
}
